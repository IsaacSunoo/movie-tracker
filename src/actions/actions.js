export const getMovies = (movies) => ({
    type: 'GET_MOVIES',
    movies
});

export const toggleFavorite = (id) => ({
  type: 'TOGGLE_FAVORITE',
  id
})

export const loginUser = (userInfo) => ({
  type: 'LOGIN_USER',
  userInfo
})