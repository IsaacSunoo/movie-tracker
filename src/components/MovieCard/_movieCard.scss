.movie-card {
    @include card-design();

    &:hover {
        transform: scale(1.05);
    }

    // grid-template-columns: repeat(3, 1fr);
    // grid-gap: 5px;
    // grid-template-areas: "image title title"
    //                      "image date date"
    //                      "image info info"
    //                      "image info info"
    //                      "image info info";

    // &:link, &:active   {
    //     text-decoration: none;
    // }
}

// .card-image {
//     grid-area: image;
//     justify-content: center;
//     margin: 10px 10px 0 0;
// }

// .movie-title {
//     grid-area: title;
// }

// .release-date {
//     grid-area: date;
// }

// .movie-info {
//     display: flex;
//     justify-content: flex-start;
//     grid-area: info;
// }