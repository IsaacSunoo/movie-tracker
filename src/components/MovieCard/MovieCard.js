import React, { Component } from 'react';
import { connect } from 'react-redux';
import { toggleFavorite } from '../../actions/actions';

class MovieCard extends Component {

    render() {
      const { poster_path, id, handleFavorite, favorites } = this.props;
      const imageSrc = 'http://image.tmdb.org/t/p/w500//' + poster_path;
      const background = { backgroundImage: `url(${imageSrc})` };

      const star = (
        <div className="star">
          {!favorites.includes(id) && <i className="fas fa-star" />}
          {favorites.includes(id) && <i className="fas fa-star active-favorite" />}
        </div>
      );
      return (
        <div className='movie-card' style={background}>
          <div onClick={() => {handleFavorite(id)}}>
            {star}
          </div>
        </div>
    )
  }
}

const mapStateToProps = (state) => ({
  favorites: state.favorites
});

const mapDispatchToProps = (dispatch) => ({
  handleFavorite: (id) => dispatch(toggleFavorite(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieCard);