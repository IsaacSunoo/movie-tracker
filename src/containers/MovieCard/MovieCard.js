import React, { Component } from 'react';
import { connect } from 'react-redux';
import { toggleFavorite } from '../../actions';
import { Link } from 'react-router-dom';
import { postFetch, deleteFetch } from '../../api';
import PropTypes from 'prop-types';

export class MovieCard extends Component {

  handleFavorite = (movieId) => {
    const { favorites, loginUser } = this.props;
    if (this.isEmpty(loginUser)) {
      console.log('NO USER LOGGED IN')
      alert('Please create account to use this feature')
    } else {
      console.log('login user if block entered');
      if (!favorites.includes(movieId)) {
        this.addFavorite();
      } else {
        this.deleteFavorite(movieId);
      }
    }
  }

  isEmpty(obj) {
    for (var key in obj) {
      if (obj.hasOwnProperty(key))
        return false;
    }
    return true;
  }

  addFavorite = async () => {
    const url = 'users/favorites/new';
    const { id, title, poster_path, release_date, vote_average, overview, loginUser, toggleFavorite } = this.props;
    const favInfo = { movie_id: id, user_id: loginUser.id, title, poster_path, release_date, vote_average, overview };
    await postFetch(url, 'POST', favInfo);
    toggleFavorite(id);
  }

  deleteFavorite = async (movieId) => {
    const { id, toggleFavorite, loginUser } = this.props;
    const url = `users/${loginUser.id}/favorites/${movieId}`;
    await deleteFetch(url);
    toggleFavorite(id);
  }

    render() {
      const { poster_path, id, favorites } = this.props;
      const imageSrc = 'http://image.tmdb.org/t/p/w500//' + poster_path;
      const background = { backgroundImage: `url(${imageSrc})` };

      const star = (
        <div className="star">
          {!favorites.includes(id) && <i className="fas fa-star" />}
          {favorites.includes(id) && <i className="fas fa-star active-favorite" />}
        </div>
      );

      return (
        <div className='movie-card' style={background}>
          <Link to={`/movie/${id}`} key={id}>
            <div className='linkdiv'>
            </div>
          </Link>
          <div onClick={() => {this.handleFavorite(id)}}>
            {star}
          </div>
        </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  loginUser: state.loginUser,
  favorites: state.favorites
});

export const mapDispatchToProps = (dispatch) => ({
  toggleFavorite: (id) => dispatch(toggleFavorite(id))
});

MovieCard.propTypes = {
  loginUser: PropTypes.object.isRequired,
  favorites: PropTypes.array.isRequired,
  toggleFavorite: PropTypes.func.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(MovieCard);