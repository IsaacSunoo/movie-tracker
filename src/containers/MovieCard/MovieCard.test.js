import React from 'react';
import { MovieCard, mapStateToProps, mapDispatchToProps } from './MovieCard';
import { shallow } from 'enzyme';
import * as actions from '../../actions';

describe('MovieCard', () => {
  let wrapper;
  let mockFavorites;
  beforeEach(() => {
    mockFavorites = [123, 444]
    wrapper = shallow(
      <MovieCard 
        favorites={mockFavorites}
      />
    )
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  
  describe('mapStateToProps', () => {
    it('should return an object with loginuser and favorites', () => {
      const mockState = {
        loginUser: {name: 'Isaac'},
        favorites: [123, 444],
        movies: [{title: 'Hey'}, {title: 'anotherMovie'}]
      }
      const expected = {
        loginUser: {name: 'Isaac'},
        favorites: [123, 444]
      }
      const result = mapStateToProps(mockState);
      expect(result).toEqual(expected)
    });
  });
  describe('mapDispatchToProps', () => {
    it('should call dispatch when using toggleFavorite', () => {
      const mockDispatch = jest.fn();
      const action = actions.toggleFavorite({name: 'Movie1'})
      const mappedDispatch = mapDispatchToProps(mockDispatch)
      mappedDispatch.toggleFavorite({name: 'Movie1'})
      expect(mockDispatch).toHaveBeenCalledWith(action)
    });
  });
  describe('handleFavorite', () => {
  
  });
  
  describe('addFavorite', () => {
  
  });
  
  describe('deleteFavorite', () => {
  
  });
});