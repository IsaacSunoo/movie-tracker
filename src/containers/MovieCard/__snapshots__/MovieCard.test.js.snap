// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MovieCard
    favorites={
      Array [
        123,
        444,
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Link
          to="/movie/undefined"
        >
          <div
            className="linkdiv"
          />
        </Link>,
        <div
          onClick={[Function]}
        >
          <div
            className="star"
          >
            <i
              className="fas fa-star"
            />
          </div>
        </div>,
      ],
      "className": "movie-card",
      "style": Object {
        "backgroundImage": "url(http://image.tmdb.org/t/p/w500//undefined)",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <div
            className="linkdiv"
          />,
          "to": "/movie/undefined",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "linkdiv",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="star"
          >
            <i
              className="fas fa-star"
            />
          </div>,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <i
                className="fas fa-star"
              />,
              false,
            ],
            "className": "star",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "fas fa-star",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
            false,
          ],
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Link
            to="/movie/undefined"
          >
            <div
              className="linkdiv"
            />
          </Link>,
          <div
            onClick={[Function]}
          >
            <div
              className="star"
            >
              <i
                className="fas fa-star"
              />
            </div>
          </div>,
        ],
        "className": "movie-card",
        "style": Object {
          "backgroundImage": "url(http://image.tmdb.org/t/p/w500//undefined)",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <div
              className="linkdiv"
            />,
            "to": "/movie/undefined",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "linkdiv",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="star"
            >
              <i
                className="fas fa-star"
              />
            </div>,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <i
                  className="fas fa-star"
                />,
                false,
              ],
              "className": "star",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "fas fa-star",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
              false,
            ],
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
